APOC NLP functions for Azure cognitive services - Failed - returned HTTP response code: 400

## Expected Behavior (Mandatory)
HTTP response: 200

## Actual Behavior (Mandatory)
HTTP response: 400

## How to Reproduce the Problem

### Simple Dataset (where it's possibile)

documentation: https://neo4j.com/labs/apoc/4.2/nlp/azure/#nlp-azure-examples-entities
currently using: neo4j 4.1.3-enterprise, APOC 4.1.0.6 and NLP 4.1.0.2
also tried with: neo4j 4.2.3-enterprise, APOC 4.2.0.1 and NLP 4.2.0.1
exactly same failure result, I have seen it work properly 4 months ago and used this code:

here some code to add test data (text for NLP entities extraction)
```
MERGE (:Article {
  uri: "https://neo4j.com/blog/pokegraph-gotta-graph-em-all/",
  body: "These days I'm rarely more than a few feet away from my Nintendo Switch and I play board games, card games and role playing games with friends at least once or twice a week. I've even organised lunch-time Mario Kart 8 tournaments between the Neo4j European offices!"
})
```

```
MERGE (:Article {
  uri: "https://en.wikipedia.org/wiki/Nintendo_Switch",
  body: "The Nintendo Switch is a video game console developed by Nintendo, released worldwide in most regions on March 3, 2017. It is a hybrid console that can be used as a home console and portable device. The Nintendo Switch was unveiled on October 20, 2016. Nintendo offers a Joy-Con Wheel, a small steering wheel-like unit that a Joy-Con can slot into, allowing it to be used for racing games such as Mario Kart 8."
})
```

here is the code to extract the entities (but fail)
```
MATCH (a:Article)
WITH collect(a) AS articles
CALL apoc.nlp.azure.entities.graph(articles, {
  key: "58abe70ff5f74497bf596cb5055f9683",
  url: "https://kevintest0702202001.cognitiveservices.azure.com",
  nodeProperty: "body",
  writeRelationshipType: "ENTITY",
  write: true
})
YIELD graph AS g
RETURN g
```
(Replaced it with the proper keys and url, feel free to try yourself if you know how ?)

included a link to this thread on Slack if you want to test more stuff or communicate ?
https://neo4j-users.slack.com/archives/C5C4JRFK7/p1613053375003700

### test machine
(only valid for a week or so)
here is the server to test on with these versions:
http://138.91.61.102:7474/browser/
username = neo4j
password = digityser

on Forum I noticed this community subject:
https://community.neo4j.com/t/apoc-nlp-functions-for-azure-cognitive-services-are-not-working/31619

I have seen it work before, don't know why all of a sudden it does not...
Also notice, that the NLP code probably uses an old API version on Azure v2.1, we are currently on v3.0 - this could be updated.
personally, I got a workaround by writing python code myself and then store in the database manually, but the APOC code is handy (if it works)

## Specifications (Mandatory)

Currently used versions

### Versions

Cores : 2 (2GHz)
Memory: 7.78 GB (24.6%)
Swap  : 8 GB
Disk  : 145 GB (45.4% ext4)
System: 18.04.1-Ubuntu

conda : 4.9.2
pip   : 21.0.1
python: 3.8.6
py2neo: 4.2.0

dbms.default_listen_address=0.0.0.0


neo4j.bloom.license_file=/plugins/bloom-plugin.license
neo4j.bloom.authorization_role=admin,architect
dbms.unmanaged_extension_classes=com.neo4j.bloom.server=/browser/bloom
dbms.tx_log.rotation.retention_policy=100M size
dbms.security.procedures.whitelist=apoc.*,gds.*
dbms.security.procedures.unrestricted=apoc.*,gds.*,bloom.*
dbms.memory.pagecache.size=2G
dbms.memory.heap.max_size=2G
dbms.directories.plugins=/plugins
dbms.directories.logs=/logs
dbms.directories.import=/import
causal_clustering.transaction_advertised_address=119fdd59aacc:6000
causal_clustering.raft_advertised_address=119fdd59aacc:7000
causal_clustering.discovery_advertised_address=119fdd59aacc:5000
apoc.import.file.enabled=true

total 58312
-rw-r--r-- 1 7474 7474 10848418 Aug 12  2020 apoc-nlp-dependencies-4.1.0.2.jar
-rw-r--r-- 1 root root 18540089 Jan 26 14:35 apoc.jar
-rw-r--r-- 1 7474 7474 11132061 Sep 30 17:49 bloom-plugin-4.x-1.4.0.jar
-rw-r--r-- 1 7474 7474       84 Feb 11 14:26 bloom-plugin.license
-rw-r--r-- 1 root root  9985742 Nov  5 16:55 graph-data-science.jar
-rw-r--r-- 1 7474 7474  9192401 Sep 30 17:49 neo4j-bloom-1.4.0-assets.zip

Entities = does not work
Key Phrases = does not work
Sentiment = Works !

```
MATCH (a:Article {uri: "https://neo4j.com/blog/pokegraph-gotta-graph-em-all/"})
CALL apoc.nlp.azure.sentiment.stream(a, {
  key: "58abe70ff5f74497bf596cb5055f9683",
  url: "https://kevintest0702202001.cognitiveservices.azure.com",
  nodeProperty: "body"
})
YIELD value
RETURN value;
```

{
  "score": 0.5,
  "id": "0"
}

## some extra information about the API on Azure for sentiment
HTTP request URL:
https://westeurope.api.cognitive.microsoft.com/text/analytics/v2.1/sentiment
```
POST https://westeurope.api.cognitive.microsoft.com/text/analytics/v2.1/sentiment HTTP/1.1
Host: westeurope.api.cognitive.microsoft.com
Content-Type: application/json
Ocp-Apim-Subscription-Key: 58abe70ff5f74497bf596cb5055f9683

{
  "documents": [
    {
      "language": "en",
      "id": "1",
      "text": "Hello world. This is some input text that I love."
    },
    {
      "language": "fr",
      "id": "2",
      "text": "Bonjour tout le monde"
    },
    {
      "language": "es",
      "id": "3",
      "text": "La carretera estaba atascada. Había mucho tráfico el día de ayer."
    }
  ]
}
```

```
Transfer-Encoding: chunked
csp-billing-usage: CognitiveServices.TextAnalytics.BatchScoring=3
x-envoy-upstream-service-time: 13
apim-request-id: ab705030-6bf8-4c94-b8c9-adbc01ce41ae
Strict-Transport-Security: max-age=31536000; includeSubDomains; preload
x-content-type-options: nosniff
Date: Thu, 11 Feb 2021 15:30:41 GMT
Content-Type: application/json; charset=utf-8

{
  "documents": [{
    "id": "1",
    "score": 0.99634093046188354
  }, {
    "id": "2",
    "score": 0.84012651443481445
  }, {
    "id": "3",
    "score": 0.334433376789093
  }],
  "errors": []
}
```

## some extra information about the API on Azure for **entities**:
HTTP request URL:
https://westeurope.api.cognitive.microsoft.com/text/analytics/v2.1/entities
```
POST https://westeurope.api.cognitive.microsoft.com/text/analytics/v2.1/entities HTTP/1.1
Host: westeurope.api.cognitive.microsoft.com
Content-Type: application/json
Ocp-Apim-Subscription-Key: 58abe70ff5f74497bf596cb5055f9683

{
  "documents": [
    {
      "language": "en",
      "id": "1",
      "text": "Hello world. This is some input text that I love."
    },
    {
      "language": "fr",
      "id": "2",
      "text": "Bonjour tout le monde"
    },
    {
      "language": "es",
      "id": "3",
      "text": "La carretera estaba atascada. Había mucho tráfico el día de ayer."
    }
  ]
}
```

```
Transfer-Encoding: chunked
csp-billing-usage: CognitiveServices.TextAnalytics.BatchScoring=3
x-envoy-upstream-service-time: 17
apim-request-id: fad4c1d6-81e7-4313-8cc3-c3afa94fd711
Strict-Transport-Security: max-age=31536000; includeSubDomains; preload
x-content-type-options: nosniff
Date: Thu, 11 Feb 2021 15:30:27 GMT
Content-Type: application/json; charset=utf-8

{
  "documents": [{
    "id": "1",
    "entities": [{
      "name": "\"Hello, World!\" program",
      "matches": [{
        "wikipediaScore": 0.2148458670347686,
        "text": "Hello world",
        "offset": 0,
        "length": 11
      }],
      "wikipediaLanguage": "en",
      "wikipediaId": "\"Hello, World!\" program",
      "wikipediaUrl": "https://en.wikipedia.org/wiki/\"Hello,_World!\"_program",
      "bingId": "7b4d3717-77ab-39ff-d8ec-1c7fd8723bd2",
      "type": "Other"
    }]
  }, {
    "id": "2",
    "entities": []
  }, {
    "id": "3",
    "entities": [{
      "name": "el día",
      "matches": [{
        "entityTypeScore": 0.8,
        "text": "el día",
        "offset": 50,
        "length": 6
      }],
      "type": "DateTime",
      "subType": "Date"
    }, {
      "name": "ayer",
      "matches": [{
        "entityTypeScore": 0.8,
        "text": "ayer",
        "offset": 60,
        "length": 4
      }],
      "type": "DateTime",
      "subType": "Date"
    }]
  }],
  "errors": []
}
```

-rwxrwxrwx 1   7474 7474 18542753 Feb 25 15:52 apoc-4.1.0.6-all.jar
-rwxrwxrwx 1   7474 7474 10848418 Feb 25 16:00 apoc-nlp-dependencies-4.1.0.2.jar
-rwxrwxrwx 1   7474 7474 11132061 Sep 30 17:49 bloom-plugin-4.x-1.4.0.jar
-rwxrwxrwx 1   7474 7474       84 Feb 11 14:26 bloom-plugin.license
-rwxrwxrwx 1   7474 7474 14522303 Feb  9 23:34 graph-data-science.jar
-rwxrwxrwx 1   7474 7474  9192401 Sep 30 17:49 neo4j-bloom-1.4.0-assets.zip

changed docker-compose:
version: '3'
services:
  neo4j_db:
    restart: always         # startup on reboot VM
    image: neo4j:4.1.3-enterprise   # Enterprise Edition
    volumes:
      - /home/ubuntu/notebooks/azuremachinelearning/neo4j/plugins:/plugins
      - /home/ubuntu/notebooks/azuremachinelearning/neo4j/logs:/logs
      - /home/ubuntu/notebooks/azuremachinelearning/neo4j/data:/var/lib/neo4j/data
      - /home/ubuntu/notebooks/azuremachinelearning/neo4j/conf:/var/lib/neo4j/conf
      - /home/ubuntu/notebooks/azuremachinelearning/neo4j/import:/import
    ports:
      - 7474:7474    # http
      - 7473:7473    # https
      - 7687:7687    # Bolt
    environment:
      - NEO4J_ACCEPT_LICENSE_AGREEMENT=yes
      - NEO4J_AUTH=neo4j/digityser
      - NEO4J_dbms_memory_pagecache_size=2G # default 512MB
      - NEO4J_dbms_memory_heap_max__size=2G # default 512MB
      - NEO4J_dbms_security_procedures_unrestricted=apoc.,gds.,bloom.*
      - NEO4J_dbms_security_procedures_whitelist=apoc.,gds.
      - NEO4J_dbms_unmanaged__extension__classes=com.neo4j.bloom.server=/browser/bloom
      - NEO4J_neo4j_bloom_license__file=/plugins/bloom-plugin.license
      - NEO4J_neo4j_bloom_authorization__role=admin,architect
      - NEO4J_apoc_import_file_enabled=true

failed to start up, it has to do with log files that needs to be cleared... try again
sudo rm neo4j/logs/*.log
sudo rm -Rf neo4j/data/databases/neo4j
sudo rm -Rf neo4j/data/transactions/neo4j

sudo docker-compose down --rmi all
sudo docker-compose up &

I fail to start Neo4j
ubuntu@myVM02:~/notebooks/azuremachinelearning/neo4j$ sudo docker-compose up &
[1] 7637
ubuntu@myVM02:~/notebooks/azuremachinelearning/neo4j$ Building with native build. Learn about native build in Compose here: https://docs.docker.com/go/compose-native-build/
Creating network "azuremachinelearning_default" with the default driver
Pulling neo4j_db (neo4j:4.1.3-enterprise)...
4.1.3-enterprise: Pulling from library/neo4j
45b42c59be33: Pull complete
a91c0c19c848: Pull complete
dbe61a45ef18: Pull complete
532e5467f958: Pull complete
e1a0ee7ccff9: Pull complete
cc1b213a4271: Pull complete
c6862d90bd9f: Pull complete
f1997d1f157d: Pull complete
Digest: sha256:2387721fb6e9933866dcc318647161a6bd1cf7f92734f996daa82bd0b0baf8f8
Status: Downloaded newer image for neo4j:4.1.3-enterprise
Creating azuremachinelearning_neo4j_db_1 ... done
Attaching to azuremachinelearning_neo4j_db_1
neo4j_db_1  | Warning: Some files inside "/data" are not writable from inside container. Changing folder owner to neo4j.
neo4j_db_1  | Changed password for user 'neo4j'.
neo4j_db_1  | Directories in use:
neo4j_db_1  |   home:         /var/lib/neo4j
neo4j_db_1  |   config:       /var/lib/neo4j/conf
neo4j_db_1  |   logs:         /logs
neo4j_db_1  |   plugins:      /plugins
neo4j_db_1  |   import:       /import
neo4j_db_1  |   data:         /var/lib/neo4j/data
neo4j_db_1  |   certificates: /var/lib/neo4j/certificates
neo4j_db_1  |   run:          /var/lib/neo4j/run
neo4j_db_1  | Starting Neo4j.
neo4j_db_1  | 2021-02-25 15:11:17.042+0000 INFO  Starting...
neo4j_db_1  | 2021-02-25 15:11:24.829+0000 INFO  ======== Neo4j 4.1.3 ========
neo4j_db_1  | 2021-02-25 15:11:32.628+0000 ERROR Failed to start Neo4j on dbms.connector.http.listen_address, a socket address. If missing port or hostname it is acquired from dbms.default_listen_address. Error starting Neo4j database server at /data/databases
neo4j_db_1  | java.lang.RuntimeException: Error starting Neo4j database server at /data/databases
neo4j_db_1  |   at org.neo4j.graphdb.facade.DatabaseManagementServiceFactory.startDatabaseServer(DatabaseManagementServiceFactory.java:198)
neo4j_db_1  |   at org.neo4j.graphdb.facade.DatabaseManagementServiceFactory.build(DatabaseManagementServiceFactory.java:158)
neo4j_db_1  |   at com.neo4j.server.enterprise.EnterpriseManagementServiceFactory.createManagementService(EnterpriseManagementServiceFactory.java:38)
neo4j_db_1  |   at com.neo4j.server.enterprise.EnterpriseBootstrapper.createNeo(EnterpriseBootstrapper.java:20)
neo4j_db_1  |   at org.neo4j.server.NeoBootstrapper.start(NeoBootstrapper.java:117)
neo4j_db_1  |   at org.neo4j.server.NeoBootstrapper.start(NeoBootstrapper.java:87)
neo4j_db_1  |   at com.neo4j.server.enterprise.EnterpriseEntryPoint.main(EnterpriseEntryPoint.java:25)
neo4j_db_1  | Caused by: org.neo4j.kernel.lifecycle.LifecycleException: Component 'com.neo4j.dbms.StandaloneDbmsReconcilerModule@497ed877' was successfully initialized, but failed to start. Please see the attached cause exception "Transaction logs contains entries with prefix 2, and the highest supported prefix is 1. This indicates that the log files originates from a newer version of neo4j.".
neo4j_db_1  |   at org.neo4j.kernel.lifecycle.LifeSupport$LifecycleInstance.start(LifeSupport.java:463)
neo4j_db_1  |   at org.neo4j.kernel.lifecycle.LifeSupport.start(LifeSupport.java:110)
neo4j_db_1  |   at org.neo4j.graphdb.facade.DatabaseManagementServiceFactory.startDatabaseServer(DatabaseManagementServiceFactory.java:189)
neo4j_db_1  |   ... 6 more
neo4j_db_1  | Caused by: org.neo4j.dbms.api.DatabaseManagementException: A triggered DbmsReconciler job failed with the following cause
neo4j_db_1  |   at com.neo4j.dbms.ReconcilerResult.join(ReconcilerResult.java:57)
neo4j_db_1  |   at com.neo4j.dbms.StandaloneDbmsReconcilerModule.startInitialDatabases(StandaloneDbmsReconcilerModule.java:95)
neo4j_db_1  |   at com.neo4j.dbms.StandaloneDbmsReconcilerModule.start(StandaloneDbmsReconcilerModule.java:85)
neo4j_db_1  |   at org.neo4j.kernel.lifecycle.LifeSupport$LifecycleInstance.start(LifeSupport.java:442)
neo4j_db_1  |   ... 8 more
neo4j_db_1  | Caused by: org.neo4j.dbms.api.DatabaseManagementException: An error occurred! Unable to start database with name `system`.
neo4j_db_1  |   at org.neo4j.dbms.database.AbstractDatabaseManager.startDatabase(AbstractDatabaseManager.java:191)
neo4j_db_1  |   at com.neo4j.dbms.database.MultiDatabaseManager.forSingleDatabase(MultiDatabaseManager.java:134)
neo4j_db_1  |   at com.neo4j.dbms.database.MultiDatabaseManager.startDatabase(MultiDatabaseManager.java:119)
neo4j_db_1  |   at com.neo4j.dbms.Transition$Prepared.doTransitionAction(Transition.java:101)
neo4j_db_1  |   at com.neo4j.dbms.Transition$Prepared.doTransition(Transition.java:88)
neo4j_db_1  |   at com.neo4j.dbms.DbmsReconciler.doTransitionStep(DbmsReconciler.java:347)
neo4j_db_1  |   at com.neo4j.dbms.DbmsReconciler.doTransitionStep(DbmsReconciler.java:348)
neo4j_db_1  |   at com.neo4j.dbms.DbmsReconciler.doTransitionStep(DbmsReconciler.java:348)
neo4j_db_1  |   at com.neo4j.dbms.DbmsReconciler.lambda$doTransitions$11(DbmsReconciler.java:316)
neo4j_db_1  |   at com.neo4j.dbms.DbmsReconciler.namedJob(DbmsReconciler.java:327)
neo4j_db_1  |   at com.neo4j.dbms.DbmsReconciler.doTransitions(DbmsReconciler.java:317)
neo4j_db_1  |   at com.neo4j.dbms.DbmsReconciler.lambda$doTransitions$9(DbmsReconciler.java:308)
neo4j_db_1  |   at java.base/java.util.concurrent.CompletableFuture$AsyncSupply.run(CompletableFuture.java:1700)
neo4j_db_1  |   at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:515)
neo4j_db_1  |   at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
neo4j_db_1  |   at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
neo4j_db_1  |   at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
neo4j_db_1  |   at java.base/java.lang.Thread.run(Thread.java:834)
neo4j_db_1  | Caused by: java.lang.RuntimeException: java.lang.RuntimeException: Error reading transaction logs, recovery not possible. To force the database to start anyway, you can specify 'unsupported.dbms.tx_log.fail_on_corrupted_log_files=false'. This will try to recover as much as possible and then truncate the corrupt part of the transaction log. Doing this means your database integrity might be compromised, please consider restoring from a consistent backup instead.
neo4j_db_1  |   at org.neo4j.kernel.database.Database.start(Database.java:496)
neo4j_db_1  |   at org.neo4j.dbms.database.AbstractDatabaseManager.startDatabase(AbstractDatabaseManager.java:187)
neo4j_db_1  |   ... 17 more
neo4j_db_1  | Caused by: java.lang.RuntimeException: Error reading transaction logs, recovery not possible. To force the database to start anyway, you can specify 'unsupported.dbms.tx_log.fail_on_corrupted_log_files=false'. This will try to recover as much as possible and then truncate the corrupt part of the transaction log. Doing this means your database integrity might be compromised, please consider restoring from a consistent backup instead.
neo4j_db_1  |   at org.neo4j.kernel.recovery.Recovery.throwUnableToCleanRecover(Recovery.java:480)
neo4j_db_1  |   at org.neo4j.kernel.recovery.LogTailScanner.findLogTail(LogTailScanner.java:168)
neo4j_db_1  |   at org.neo4j.kernel.recovery.LogTailScanner.getTailInformation(LogTailScanner.java:342)
neo4j_db_1  |   at org.neo4j.kernel.recovery.Recovery.validateStoreId(Recovery.java:401)
neo4j_db_1  |   at org.neo4j.kernel.database.Database.checkStoreId(Database.java:534)
neo4j_db_1  |   at org.neo4j.kernel.database.Database.validateStoreAndTxLogs(Database.java:513)
neo4j_db_1  |   at org.neo4j.kernel.database.Database.start(Database.java:386)
neo4j_db_1  |   ... 18 more
neo4j_db_1  | Caused by: org.neo4j.kernel.impl.transaction.log.entry.UnsupportedLogVersionException: Transaction logs contains entries with prefix 2, and the highest supported prefix is 1. This indicates that the log files originates from a newer version of neo4j.
neo4j_db_1  |   at org.neo4j.kernel.impl.transaction.log.entry.LogEntryVersion.select(LogEntryVersion.java:93)
neo4j_db_1  |   at org.neo4j.kernel.impl.transaction.log.entry.VersionAwareLogEntryReader.readLogEntry(VersionAwareLogEntryReader.java:89)
neo4j_db_1  |   at org.neo4j.kernel.impl.transaction.log.LogEntryCursor.next(LogEntryCursor.java:53)
neo4j_db_1  |   at org.neo4j.kernel.recovery.LogTailScanner.findLogTail(LogTailScanner.java:123)
neo4j_db_1  |   ... 23 more
neo4j_db_1  | 2021-02-25 15:11:32.654+0000 INFO  Neo4j Server shutdown initiated by request

--- a/full/src/main/kotlin/apoc/nlp/azure/azure/RealAzureClient.kt
+++ b/full/src/main/kotlin/apoc/nlp/azure/azure/RealAzureClient.kt
@@ -42,7 +42,7 @@ class RealAzureClient(private val baseUrl: String, private val key: String, priv
         connection.setRequestProperty("Ocp-Apim-Subscription-Key", subscriptionKeyValue)
         connection.doOutput = true
 
-        connection.setRequestProperty("Content-Type", "text/json")
+        connection.setRequestProperty("Content-Type", "application/json")
         DataOutputStream(connection.outputStream).use { it.write(JsonUtil.writeValueAsBytes(mapOf("documents" to convertInput(data)))) }
 
         return connection.inputStream

